src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,154 @@
+            }
+            Node<T> newNode = new Node<>(value);
+            newNode.next = current.next;
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
+    }
+
+    public void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = current.data;
+        current.data = value;
+        return oldValue;
+    }
+
+    public T unlink(Node<T> node) {
+        T removedValue = node.data;
+        if (node == head) {
+            head = node.next;
+        } else {
+            Node<T> prev = head;
+            while (prev != null && prev.next != node) {
+                prev = prev.next;
+            }
+            if (prev != null) {
+                prev.next = node.next;
+            }
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue;
+        if (index == 0) {
+            removedValue = head.data;
+            head = head.next;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            removedValue = current.next.data;
+            current.next = current.next.next;
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        Node<T> prev = null;
+        while (current != null) {
+            if ((current.data == null && object == null)
+                    || (current.data != null && current.data.equals(object))) {
+                if (prev == null) {
+                    head = current.next;
+                } else {
+                    prev.next = current.next;
+                }
+                size--;
+                return true;
+            }
+            prev = current;
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T data;
+        private Node<T> next;
+
```

Comment:
```suggestion
    private static class Node<T> {
        private T data;
        private Node<T> next;
        private Node<T> prev;
```
your linked list should be bidirectional